src/plugins/lua/contextal.lua

```diff
@@ -0,0 +1,319 @@
+    if parse_err then
+      rspamd_logger.err(task, 'cannot parse JSON: %s', err)
+      maybe_defer(task)
+      return
+    end
+    local obj = parser:get_object()
+    local work_id = obj.work_id
+    if work_id then
+      task:get_mempool():set_variable('contextal_work_id', work_id)
+    end
+    task:insert_result(settings.submission_symbol, 1.0,
+        string.format('work_id=%s', work_id or 'nil'))
+  end
+
+  local req = {
+    object_data = {['data'] = task:get_content()},
+  }
+  if settings.request_ttl then
+    req.ttl = {['data'] = tostring(settings.request_ttl)}
+  end
+  if settings.max_recursion then
+    req.maxrec = {['data'] = tostring(settings.max_recursion)}
+  end
+  rspamd_http.request({
+      task = task,
+      url = settings.submit_url,
+      body = lua_util.table_to_multipart_body(req, static_boundary),
+      callback = http_callback,
+      headers = {
+        ['Content-Type'] = string.format('multipart/form-data; boundary="%s"', static_boundary)
+      },
+      timeout = settings.http_timeout,
+      gzip = settings.gzip,
+      keepalive = settings.keepalive,
+      no_ssl_verify = settings.no_ssl_verify,
+  })
+end
+
+local function cache_hit(task, err, data)
+  if err then
+    rspamd_logger.err(task, 'error getting cache: %s', err)
+  else
+    process_cached(task, data)
+  end
+end
+
+local function submit_cb(task)
+  if cache_context then
+    redis_cache.cache_get(task,
+        task:get_digest(),
+        cache_context,
+        settings.cache_timeout,
+        submit,
+        cache_hit
+    )
+  else
+    submit(task)
+  end
+end
+
+local function action_cb(task)
+  local work_id = task:get_mempool():get_variable('contextal_work_id', 'string')
+  if not work_id then
+    rspamd_logger.err(task, 'no work id found in mempool')
+    return
+  end
+
+  local function http_callback(err, code, body, hdrs)
+    if err then
+      rspamd_logger.err(task, 'http error: %s', err)
+      maybe_defer(task)
+      return
+    end
+    if code ~= 200 then
+      rspamd_logger.err(task, 'bad http code: %s', code)
+      maybe_defer(task)
+      return
+    end
+    local parser = ucl.parser()
+    local _, parse_err = parser:parse_string(body)
+    if parse_err then
+      rspamd_logger.err(task, 'cannot parse JSON: %s', err)
+      maybe_defer(task)
+      return
+    end
+    local obj = parser:get_object()
+    return process_actions(task, obj, false)
+  end
+
+  rspamd_http.request({
+      task = task,
+      url = settings.actions_url .. work_id,
+      callback = http_callback,
+      timeout = settings.http_timeout,
+      gzip = settings.gzip,
+      keepalive = settings.keepalive,
+      no_ssl_verify = settings.no_ssl_verify,
+  })
+end
+
+local function set_url_path(base, path)
+  local slash = base:sub(#base) == '/' and '' or '/'
+  return base .. slash .. path
+end
+
+settings = lua_util.override_defaults(settings, opts)
+
+local res, err = config_schema:transform(settings)
```

Comment:
Shouldn't we enrich it with Redis parameters + caching schema as we use caching toolkit?

Reply:
Yes, probably so; I'll check that

Reply:
Well, the schema seems to be in there somehow; also I don't see that `lua_cache` exports a schema

Reply:
I think it should, but let it be there for now. In general, the configuration dependent tool (like lua_cache) should export some `enrich_schema` method like `lua_redis` does.